docs/resource_management.md

```diff
@@ -0,0 +1,68 @@
+# Resource management
+
+This document describes the fundamentals of resource management in permissions-api.
+
+## Resource types
+
+Every resource in permissions-api has a type that defines the actions and relationships that can be scoped to that resource. By default, the only two resource types that exist in permissions-api are _role_ and _tenant_, which define sets of allowed actions and operational context respectively and are necessary for the service to function.
+
+## Resource lifecycle events
+
+permissions-api consumes resource lifecycle events over [NATS][nats]. This section describes the subjects and message formats the service expects.
+
+[nats]: https://nats.io
+
+### Message format
+
+permission-api expects lifecycle event messages in the format described in [`go.infratographer.com/x/pubsubx`][pubsubx], and interprets the message fields as follows:
+
+* `subject_urn`: The URN of the resource the event is about
+* `event_type`: The type of lifecyle event for the resouce. Must match one of the defined lifecycle events
+* `additional_subjects`: Resources related to the resource. These map to a relationship with both a relation and resource type that match the resource type in the URN
+
+[pubsubx]: https://github.com/infratographer/x/blob/v0.0.7/pubsubx/message.go
+
+### Subjects
+
+When listening for events, permissions-api subscribes to subjects of the form `{namespace}.{resource_type}.{lifecycle_event}`, where `namespace` is the configured namespace for resource types, `resource_type` is any resource type not directly managed by permissions-api (such as `role`), and `lifecycle_event` is one of the following defined lifecycle events.
+
+### Lifecycle events
+
+The following lifecycle events are recognized by permissions-api and will result in the described effects.
+
+#### `create`
+
+A `create` event signals that a given resource has been created. All relationships described in the given message will be created. It is an error to send multiple `create` events for a single resource.
+
+#### `update`
+
+An `update` event signals that a given resource has been updated. All relationships described in the given event message will be created if they do not already exist, and all other relationships not defined in the message will be deleted. Sending multiple `update` events for the same resource is permitted.
+
+#### `delete`
+
+A `delete` event signals that a given resource has been deleted. All relationships will be deleted where the resource in the relationship (as opposed to the subject) matches the provided resource in the message. Note that this means resource deletion in permissions-api will not result in cascading deletes of relationships. It is the responsibility of all services that manage resources to send explicit `delete` events for every resource they manage.
```

Comment:
Although this keeps the permissions-api simpler, I'm not convinced we dont want to delete all relationships with a resource that's deleted. For our current use case, this may be okay.